parlai/core/teachers.py

```diff
@@ -2539,11 +2547,23 @@ def get_chunk(self):
         # abstract method `load_from_chunk` returns a list of tuples
         output = self.load_from_chunk(next_chunk)
 
-        if self.is_train:
+        if DatatypeHelper.should_shuffle(self.datatype):
             # randomize the samples
             random.Random().shuffle(output)
         return output, chunk_reset_cnt
 
+    def next_example(self):
+        # next_example will always fail to provide useful signal on whether
+        # we're at the end of an epoch in chunk teacher. Instead, the queue
+        # empties and we simply start outputting pads forever. As such, we'll
+        # measure epochs when we start receiving only pads.
+
+        # (This isn't relevant for the training loop, which loops for ever and
+        # never "epochs").
+        retval, fake_epoch_done = super().next_example()
+        real_epoch_done = retval.is_padding()
```

Comment:
I assume this works OK with dynamic batch padding, since that occurs at a level higher than the teacher?

Reply:
Yes, this has been tested with all variants of batching and ensured that all examples are seen.